server/src/main/java/io/crate/protocols/postgres/DelayableWriteChannel.java

```diff
@@ -0,0 +1,326 @@
+        return delegate.voidPromise();
+    }
+
+    @Override
+    public int compareTo(Channel o) {
+        return delegate.compareTo(o);
+    }
+
+    @Override
+    public ChannelId id() {
+        return delegate.id();
+    }
+
+    @Override
+    public EventLoop eventLoop() {
+        return delegate.eventLoop();
+    }
+
+    @Override
+    public Channel parent() {
+        return delegate.parent();
+    }
+
+    @Override
+    public ChannelConfig config() {
+        return delegate.config();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public boolean isRegistered() {
+        return delegate.isRegistered();
+    }
+
+    @Override
+    public boolean isActive() {
+        return delegate.isActive();
+    }
+
+    @Override
+    public ChannelMetadata metadata() {
+        return delegate.metadata();
+    }
+
+    @Override
+    public SocketAddress localAddress() {
+        return delegate.localAddress();
+    }
+
+    @Override
+    public SocketAddress remoteAddress() {
+        return delegate.remoteAddress();
+    }
+
+    @Override
+    public ChannelFuture closeFuture() {
+        return delegate.closeFuture();
+    }
+
+    @Override
+    public boolean isWritable() {
+        return delegate.isWritable();
+    }
+
+    @Override
+    public long bytesBeforeUnwritable() {
+        return delegate.bytesBeforeUnwritable();
+    }
+
+    @Override
+    public long bytesBeforeWritable() {
+        return delegate.bytesBeforeWritable();
+    }
+
+    @Override
+    public Unsafe unsafe() {
+        return delegate.unsafe();
+    }
+
+    @Override
+    public ChannelPipeline pipeline() {
+        return delegate.pipeline();
+    }
+
+    @Override
+    public ByteBufAllocator alloc() {
+        return delegate.alloc();
+    }
+
+    @Override
+    public Channel read() {
+        return delegate.read();
+    }
+
+    @Override
+    public Channel flush() {
+        return delegate.flush();
+    }
+
+    public Channel bypassDelay() {
+        return delegate;
+    }
+
+    public void delayWritesUntil(CompletableFuture<?> future) {
+        DelayedWrites delayedWrites;
+        if (delay == null) {
+            delayedWrites = new DelayedWrites(future);
+        } else {
+            delayedWrites = new DelayedWrites(delay.future.thenCompose(ignored -> future));
+        }
+        this.delay = delayedWrites;
+        future.whenComplete((res, err) -> {
+            delayedWrites.runDelayed();
+            if (delay == delayedWrites) {
+                delay = null;
+            }
+        });
+    }
+
+    static class DelayedWrites {
+
+        private final ArrayDeque<Runnable> delayed = new ArrayDeque<>();
+        private final CompletableFuture<?> future;
+
+        public DelayedWrites(CompletableFuture<?> future) {
+            this.future = future;
+        }
+
+        public void add(Runnable runnable) {
+            delayed.add(runnable);
```

Comment:
As `ArrayDeque` isn't thread-safe, don't we have to ensure thread-safety here or on the caller side?
(same on the below `daylayed.poll()`)